Fix #11267: Added audit job to determine the limit of choices length.

SAEb-ai opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #11267.
2. This PR does the following: It adds an audit job to determine the limit of choices length in Item Selection Input Interaction or Multiple Choice Input Interaction.

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

SAEb-ai committed:
Added Audit job

SAEb-ai committed:
Added test for audit job

SAEb-ai committed:
Added job to job_registry.py

SAEb-ai committed:
Fixed minor and lint issues

SAEb-ai committed:
Fixed Lint Issues

vojtechjelinek commented on core/domain/interaction_jobs_one_off.py:
```diff
@@ -152,6 +152,43 @@ def reduce(key, values):
 
+class MultipleItemInteractionLtOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces total number of all (exploration, state) pairs
+    that use the Multiple choice interaction or Item Selection
+    Interaction and total number of all (exploration, state) pairs
+    that use the Multiple choice interaction or Item selection
+    Interaction whose choice length is greater than 30.
+    """
```
This is not true you do not return the exploration and state pairs. You return exploration ID and the length of the choice.

SAEb-ai replied:
Done

vojtechjelinek commented on core/domain/interaction_jobs_one_off_test.py:
```diff
@@ -755,6 +755,224 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+            job_id)
+        self.process_and_flush_pending_mapreduce_tasks()
+
+        actual_output = (
+            interaction_jobs_one_off
+            .MultipleItemInteractionLtOneOffJob.get_output(job_id))
+        expected_output = (2, 0)
+        self.assertEqual(actual_output, expected_output)
```
Why is the output so confusing?

vojtechjelinek replied:
I would expect something like `('SUCCESS', 2)` or similar.

SAEb-ai replied:
Done

vojtechjelinek commented on core/domain/interaction_jobs_one_off_test.py:
```diff
@@ -755,6 +755,224 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+        # Start MultipleItemInputInteractionLtOneOffJob job
+        # on sample exploration whose interaction is
+        # Item Selection Input.
+        job_id = (
+            interaction_jobs_one_off
+            .MultipleItemInteractionLtOneOffJob.
+            create_new())
+        interaction_jobs_one_off.MultipleInputInteractionLtOneOffJob.enqueue(
+            job_id)
+        self.process_and_flush_pending_mapreduce_tasks()
+
+        actual_output = (
+            interaction_jobs_one_off
+            .MultipleInputInteractionLtOneOffJob.get_output(job_id))
+        expected_output = (2, 0)
+        self.assertEqual(actual_output, expected_output)
```
Can you put this part of code to one common method that will be used by all your tests? Similart to how we do this in other tests for jobs.

SAEb-ai replied:
Done

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

SAEb-ai committed:
Addresed Review Comments

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai commented on core/domain/interaction_jobs_one_off_test.py:
```diff
@@ -755,6 +763,205 @@ def test_no_action_is_performed_for_deleted_exploration(self):
+        actual_output = (
+            interaction_jobs_one_off
+            .MultipleInputInteractionLtOneOffJob.get_output(job_id))
+        expected_output = ('LONGER THAN 30', 38)
```
@vojtechjelinek Can you please help me how to write the expected output in unicode because I am a little weak in that.
`('LONGER THAN 30', (item.id, choice_length))` .I think this is the code which would come to the expected output.
Thanks!

vojtechjelinek replied:
I think if you rework the run method according to https://github.com/oppia/oppia/blob/develop/core/domain/activity_jobs_one_off_test.py#L76 you should be able to compare the exact output.

SAEb-ai commented:
@vojtechjelinek PTAL.
I have addressed your previous review comments and also left one comment from myself.
Thanks.

vojtechjelinek commented on core/domain/interaction_jobs_one_off_test.py:
```diff
@@ -48,6 +48,14 @@ def mock_validate(unused_self):
     pass
 
 
+def start_multiple_item_interaction_lt_one_off_job(self, job_class):
+    """Helper function to start MultipleItemInteractionLtOneOffJob."""
+    job_id = job_class.create_new()
+    job_class.enqueue(job_id)
+    self.process_and_flush_pending_mapreduce_tasks()
+    return job_id
+
```
Put this function to the `MultipleItemInteractionLtOneOffJobTests` and mark it as private inside (prefix the function name with `_`).

vojtechjelinek replied:
Btw it would be useful to rework this to be similar to https://github.com/oppia/oppia/blob/develop/core/domain/activity_jobs_one_off_test.py#L76

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

SAEb-ai committed:
Merge remote-tracking branch 'upstream/develop' into limit-job

DubeySandeep committed:
Revert "Fixes part of #8472: Migrate Exploration Rights Service (#11974)" (#12305)

This reverts commit 7e4a92345dce48dce7a3066c5697f3aceb08b667.

seanlip committed:
Remove old conversion functions and upgrade all exploration demo yaml files to the latest schema versions. (#12256)

* Upgrade all exploration demo yaml files to the latest schema versions.

* Delete a subset of existing upgrade functionality.

* More fixes.

* More fixes.

* Fix lint issues.

* Fix test failures.

* Fix small error.

* Fix failures in editor_test.py.

* Fix last test error.

* Address some issues with test coverage.

aishwary023 committed:
Fix part of #9749: Migrate admin-navbar and admin-dev-mode-activities-tab & added api service for admin-navbar (#12189)

* Migrated admin-navbar and admin-dev-mode-activities-tab

* Made css changes

* Migrated spec file

* Fix lint

* Refactoed admin-backend.api.service to add admin-dev-activities-tab-component service

* Lint fix

* Added png sanitizer and refactored svg-sanitizer.service to img-sanitizer.service

* lint fix

* fix comment punctutaion

* Added test for admin-backend-api-service.ts

* fixed error

* Added unit tests for admin-dev-mode-activities component

* Removed admin-dev-mode-activities-tab comonent from check_frontend_coverage file

* Fix lint checks

* Addressed review comments

* Lint fix

* Fixed failing e2e tests

* Changes img-sanitization to image-sanitization

* Fixed lint issues

* Fixed lint error

* Addressed review comments

* Fixed lint errors

* Fixed lint errors

* Fixed spec file of admin-navbar

* Fixed spec file of admin-navbar

* Addressed review comments

* Update core/templates/services/svg-sanitizer.service.spec.ts

Co-authored-by: Srijan Reddy <srijan@srijanreddy.com>

* Reviewed comments

* Addressed review comments

* Lint fix

* Fixed coverage

* Addressed reviews

* Fixed lint errors

Co-authored-by: Srijan Reddy <srijan@srijanreddy.com>

brianrodri committed:
Revert "Fix #11462: Migrate to Firebase authentication (#12172)" (#12313)

This reverts commit d81b16e5896e5373c16abd9acd8c153a7e995f09.

ashutoshc8101 committed:
Fix part of #9749: Migrates image-uploader.directive.ts to angular (#12128)

* image-uploader.directive.ts migrated

* linting issues

* requested changes

* lints

* e2e fixed

ARKEOLOGIST committed:
Fix #12252 : Make search bar in classroom page more responsive for mobile screens (#12289)

vojtechjelinek committed:
Add permission decorator for primary admin (#12283)

ashutoshc8101 committed:
Fix part of #9749: Admin jobs tab component migrated (#12076)

* admin-config-tab migrated

* fixed identation

* minor fixes

* request-changes

* linting issues

* linitng issues

* Changed backend api service

* requested changes

* nits

* lint

* lint

ARKEOLOGIST committed:
Fix #12260 : Adds media queries to prevent overflow of tabs in question editor (#12281)

ARKEOLOGIST committed:
Fix part of #12253 : Added z-index styles to ensure dropdown does not overlap "Coming Soon" label (#12288)

aasiffaizal committed:
1. Added async for functions which returns promise (#12151)

2. Renamed async functions by adding Async to the tail of function name.

Co-authored-by: Aasif Faizal <aasif.faizal@paytmmall.com>

FareesHussain committed:
Fix part of #11475: Migrate _validate_post_commit_is_private to Apache Beam (#12272)

* created ValidatePostCommitIsPrivate && ModelInvalidCommitStatus

* Added tests

* Fix lints

* fix code coverage

* Resolving Requested changes

* Removing extra lines

* Removing the subclass for BaseCommitLogEntryModel

* Removing the subclass for BaseCommitLogEntryModel

* Update jobs/base_model_validator.py

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

ReshuKumari committed:
Fix #12178: Make questions appear on practice session (#12295)

* fixing retry test button

* addressing review

vojtechjelinek committed:
Refactor of the user services and android constants (#12229)

aasiffaizal committed:
Fix part of #10306: Add "async" keyword to asynchronous functions that return a promise  (#12302)

lkbhitesh07 committed:
Fix part of #7450 : Modified private methods in "jobs_test.ContinuousComputationTests" (#12133)

EricZLou committed:
Additional Classroom Page Fixes (change "Chapters" to "Lessons") (#12309)

* change chapters to lessons

* added comment

kishanhitk committed:
Fix part of #10798 : End-to-End tests for profileMenuFlow.js  (#12082)

Rijuta-s committed:
Autofocus editors:Adds autofocus to topic,skill and exploration editor. (#12089)

* adding autofocus to exploration editors

* adding autofocus to all tabs of exploration editor

* adding autofocusto skill and topic editors

* adding autofocusto skill and topic editors

* minor fixes for lint

* fixing lint issues

* minor lint fixes

* minor lint fixes

* adding test to state-name-editor

* minor changes

* minor fixes for topic dashboard

* fixing lint issues

* adding unit test to the files

* fixing lint issues

* fixing minor issues

* fixing lint

* fixing unit tests

* adding spec file to exploration-title editor

* minor fix

* fixing spec file in exploration-title-editor

* fixing lint issues

* minor fix

* fixing e2e test

* minor fix

* fixing lint issues

* adding frontend tests

* minor fix

* fixing frontend tests

* fixing frontend tests

* fix linting

* applying suggestion from code owners

* fixing lint issues

* fixing front end test

* Update core/templates/pages/topic-editor-page/questions-tab/topic-questions-tab.directive.ts

Co-authored-by: Kevin Thomas <kevintab@tutanota.com>

* applying suggestions

* fixing lint issues

* fixing front end tests

* fixing front end tests

* fixing lint issues

* minor fix

* Adding comment to clarify $timeout use

Co-authored-by: Kevin Thomas <kevintab@tutanota.com>

rohitkatlaa committed:
Email dashboard - Send emails by custom predicates (#12041)

U8NWXD committed:
Set a name for CI checks on custom eslint checks (#12325)

The name allows this test to be identified when configuring branch protection rules.

Commit:
Revert "Fix part of #9749: Migrates image-uploader.directive.ts to angular (#12128)" (#12335)

This reverts commit 87857aad2363d67d088d8eda03352b52c96fc4f9.

Naman-1234 committed:
Added async keyword to synchronous function that return a promise (#12291)

Commit:
Fix #12224: Edit question flow fixes (#12321)

* fix question edit flow issues

* fix lint checks

* remove extra comma

Radesh-kumar committed:
Fix #12206: Modal will be opened after clicking clock icon (#12331)

atpug22 committed:
Audit Job to check for explorations with number of ratio terms greater than 10 (#12157)

* commit

* wrote audit job

* completed audit job

* removed the solution for issue

* fixed lint issues

* fixed lint issues

* added success case in audit job

* lint issue fixed

* removed UTF-8

* Update core/domain/exp_jobs_one_off.py

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

* resolved conflicts

Co-authored-by: Vojtěch Jelínek <vojtin.j@gmail.com>

brianrodri committed:
Fixes part of #11462: Update install_backend_python_libs to respect direct urls from GitHub (#12315)

* initial commit

* add test requirements

* rename test file

ARKEOLOGIST committed:
Fix #12306: Changes the order of headings on the Splash Page (#12330)

* Changing the appearance of text on splash page

* Making CSS changes to make page look same

* Changing student name and quotes to make tags semantically correct

aasiffaizal committed:
Fix part of #11776. Adding/Updating 'Args' and 'Return' in docstrings (#12287)

* Added/updated args, return and raises in docstring

* Fixed lint issues

* Removed unwanted Raises

* Update python_utils.py

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

* Update utils.py

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

* Update utils.py

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

* Review changes

* fixed linting issue

* fixed linting issue

* Replaced more exceptions with validation errors

* refactor

Co-authored-by: Nithesh N. Hariharan <nithusha21@yahoo.co.in>

Commit:
Show skillreview RTE component when submitting a question via the community dashboard (#12340)

* show skillreview RTE component when submitting a question (community dashboard)

* rename methods to 'contributor' dashboard

nisargsc committed:
Fix part of #10474: Make typescript checks strict in NumericInputRulesService (#12320)

gp201 committed:
Update to the Proof that changes are correct section. (#12077)

gp201 committed:
Fix #11333: Correctly enable publish and transfer to community buttons in settings tab of exploration editor (#11690)

Rijuta-s committed:
Fix part of #11464: Fixes task 8 by allowing user to edit skill difficulty . (#12308)

* adding question editor modal

* adding question difficulty modal

* fixing lint issues

* adding test for  question-suggestion-editor-modal to open question difficulty modal

* minor fix

* fixing lint issues

* fixing frontend tests

* fixing lint issue

* Update question-suggestion-editor-modal.directive.html

Fixing e2e test

vojtechjelinek committed:
Update authors and changelog for v3.0.9 (#12357)

* Update CHANGELOG

* Update AUTHORS

* Update CONTRIBUTORS

* Update core/templates/pages/about-page/about-page.constants.ts

* Update package.json

mridul-netizen committed:
Fix part of #8016: Added a separate backend api service for learner-view-info-backend-api.service.ts (#12132)

* added separate backend service for learner-view-info

* updated

* updated

* removed async

* updated

* updated

darksun27 committed:
Leave: Till 10th April (#12329)

Temporary Change

jcqli committed:
Fix part of #9749: Migrate skill selector to angular (#11791)

* Fixes Part of #8472: Migrate Question Player Enginer Service to Angular 8 (#11285)

* Migrate question player enginer service to Angular 8

* Syntax error

* Lint fix

* Minor fix

* Move setting question player context from constructor to init function

* Begin initial controller migration, removing $scope

* Initial migration of HTML

* Add migrated component to shared-module

* Change directive to component

* Fix linter errors

* [skip ci] Migrate skill-selector declaration and interpolation in html

* [skip ci] More interpolation?

* Rename component to select-skill

* [skip ci] Use event emitter for categorized skills

* [skip ci] Use angular material components and local vars in component controller

* [skip ci] Subscribe to event emitter in modal?

* [skip ci] Rename select-skill to skill-selector

* [skip ci] Address review comments

* [ci skip] add types to skill-selector bindings

* [skip ci] Create pipe for filtering serach

* [skip ci] Declare let in ngFor

* [skip ci] Use angularjs filter for serach

* [skip ci] Use for...of for mat elements

* [skip ci] uUse keyvalue pipe for iterating dicts

* [skip ci] Cleanup comments

* [skip ci] Add test for pipe

* Fixing naming errors with skill-selector

* [skip ci] Remove formatting error

* [skip ci] Fix linter errors

* [skip ci] Add MaterialModule to SharedComponent

* [skip ci] Rename pipe filter for clarity

* [skip ci] Fix typo

* [skip ci] Use pipe filter in controller

* [skip ci] Adjust style

* [skip ci] Fix typo with binding

* [skip ci] Remove fdescribe

* Inject pipe in root, fix formatting

* [skip ci] Use local var for currently selected categorized skills

* Remove *ngIf for subtopic filters

* Add more filter pipe tests

* Add more pipe filter tests

* Fix typo

* Address review comments

* Fix typoe

* [skip ci] Change name to oppia-skill-selector

* Apply review comments

* Remove space, push for ci checks again

* Remove impossible case

* Add test coverage for selecting a skill

* Make list alphabetical

* [skip ci] Apply review comments

* Fix linter error

* Set selected skill in merge skill controller

* Check null untriaged skills

Co-authored-by: Prakash Subedi <4537961+subedi90@users.noreply.github.com>

DubeySandeep committed:
Revert "Fixes part of #11462: Update install_backend_python_libs to respect direct urls from GitHub (#12315)" (#12363)

This reverts commit ff7421ee424955fc86b1a96012965165cd41be12.

SAEb-ai committed:
Merge branch 'limit-job' of https://github.com/SAEb-ai/oppia into limit-job

SAEb-ai committed:
Address review comments

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed Lint Issues

SAEb-ai committed:
Fixed another set of lint issues

SAEb-ai committed:
Fixed another set of lint issues
